### PR TITLE
IGNITE-19119 Prepare org.apache.ignite.internal.storage.index.IndexStorage to built index

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-19119

### CODE REVIEW

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/index/IndexStorage.java [hunk 1]
```diff
@@ -49,9 +51,29 @@ public interface IndexStorage {
...
     void remove(IndexRow row) throws StorageException;
+
+    /**
+     * Returns the row ID for which the index needs to be built, {@code null} means that the index building has completed.
+     *
+     * <p>If index building has not started yet, it will return {@link RowId#lowestRowId(int)}.
+     *
+     * @throws StorageException If failed to get the row ID.
+     */
+    @Nullable RowId getNextRowIdToBuild();
```
Comment (sashapolo) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/index/IndexStorage.java:67]: All other methods in this class declare `throws StorageException` explicitly

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorage.java [hunk 2]
```diff
@@ -1001,4 +983,8 @@ rowId, rowTimestamp, createStorageInfo()
 
         return removeWriteOnGc.getResult();
     }
+
+    IndexMeta createIndexMetaForNewIndex(UUID indexId) {
```
Comment (sashapolo) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorage.java:987]: should it be private?

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbUtils.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.UUID;
+import org.apache.ignite.internal.storage.RowId;
+
+/**
+ * Helper class for storages based on RocksDB.
+ */
+public class RocksDbUtils {
```
Comment (sashapolo) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbUtils.java:28]: We already have a `RocksUtils` class, let's rename this one to something like `StorageUtils`

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/IndexMetaKey.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+package org.apache.ignite.internal.storage.pagememory.index.meta;
+
+import java.util.UUID;
+
+/**
+ * Index meta key.
+ */
+public class IndexMetaKey {
+    protected final UUID indexId;
```
Comment (sashapolo) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/IndexMetaKey.java:26]: Why is this field `protected` if it has a public getter?

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/index/IndexStorage.java (hunk 1 above)
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/index/IndexStorage.java:67]: But this exception is not checked, there is no need to specify it, documentation is enough.
I can remove it in other methods, but I think you should not change it.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorage.java (hunk 2 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorage.java:987]: No, it is used in successors.

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbUtils.java (hunk 3 above)
Comment (tkalkirill) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbUtils.java:28]: `StorageUtils` already exists, I'll try to rename it to `RocksDbStorageUtils`.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/IndexMetaKey.java (hunk 4 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/IndexMetaKey.java:26]: I don't see anything bad in this.

Comment (sashapolo) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/IndexMetaKey.java:26]: oh come on, I thought we sorted it out the last time. Why do we need protected fields, it just confuses everyone and provides absolutely no benefits

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbUtils.java (hunk 3 above)
Comment (sashapolo) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbUtils.java]: sure

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/IndexMetaKey.java (hunk 4 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/IndexMetaKey.java:26]: I do not agree with you, but I'm just too lazy to argue with you, this is not a significant problem.

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/index/IndexStorage.java (hunk 1 above)
Comment (ibessonov) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/index/IndexStorage.java:67]: From my side - exception in the signature simply makes it more explicit.
StorageException should have been checked, and we wouldn't have this discussion in such case.

Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/index/IndexStorage.java:67]: We discussed it personally, decided to remake `StorageException` into a checked in the future and therefore add it to the signature.

Comment (sashapolo) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/index/IndexStorage.java]: Please don't do that, checked exceptions suck and should not be used

**Review by sashapolo — APPROVED** [review]

**Pull request merged**